library/src/main/java/com/google/android/exoplayer/extractor/mp4/AtomParsers.java

```diff
@@ -1154,4 +1154,119 @@ public AvcCData(List<byte[]> initializationData, int nalUnitLengthFieldLength,
+      final Atom.LeafAtom stz2Atom = stblAtom.getLeafAtomOfType(Atom.TYPE_stz2);
+
+      if (stszAtom == null && stz2Atom == null) {
+        throw new ParserException("Track has no sample table size information");
+      }
+
+    return stszAtom != null ? new StszSampleSizeBox(stszAtom) : new Stz2SampleSizeBox(stz2Atom);
+  }
+
+  /**
+   * Interface that allows us to abstract away the various implementations of sample size boxes. (IE: stsz and stz2)
+   */
+  interface SampleSizeBox {
+    /**
+     * @return the number of samples in this atom.
+     */
+    int getSampleCount();
+
+    /**
+     * @return the fixed sample size read from the atom. If 0, the atom has non-fixed sample sizes.
+     */
+    int getFixedSampleSize();
+
+    /**
+     * @param i the index in the sample size array we wish to get the size from.
+     * @return the size for the sample at the index passed.
+     */
+    int getSampleSizeAt(int i);
+  }
+
+  /**
+   * Reads sample sizes out of a 'stsz' atom.
+   */
+  static final class StszSampleSizeBox implements SampleSizeBox {
+    public final Atom.LeafAtom atom;
+    public final int fixedSampleSize;
+    public final int sampleCount;
+    public StszSampleSizeBox(Atom.LeafAtom stszAtom) {
+      atom = stszAtom;
+      atom.data.setPosition(Atom.FULL_HEADER_SIZE);
+      fixedSampleSize = atom.data.readUnsignedIntToInt();
+      sampleCount = atom.data.readUnsignedIntToInt();
+    }
+
+    @Override
+    public int getSampleCount() {
+      return sampleCount;
+    }
+
+    @Override
+    public int getFixedSampleSize() {
+      return fixedSampleSize;
+    }
+
+    @Override
+    public int getSampleSizeAt(int i) {
+      return atom.data.readUnsignedIntToInt() ;
+    }
+  }
+
+  /**
+   * Reads samples out of a 'stz2' atom.
+   */
+  static final class Stz2SampleSizeBox implements SampleSizeBox {
+    private static final int FIRST_DATA_POSITION = Atom.FULL_HEADER_SIZE + 8;
+    public final Atom.LeafAtom atom;
+    public final int sampleCount;
+    public final int fieldSize;  // 4, 8, or 16.
+    public Stz2SampleSizeBox(Atom.LeafAtom stz2Atom) {
+      atom = stz2Atom;
+      atom.data.setPosition(Atom.FULL_HEADER_SIZE);
+      // mask out the reserved bits (24) and read just the field_size bits (8).
+      fieldSize = atom.data.readUnsignedIntToInt() & 0x000000FF;
+      sampleCount = atom.data.readUnsignedIntToInt();
+    }
+
+    @Override
+    public int getSampleCount() {
+      return sampleCount;
+    }
+
+    @Override
+    public int getFixedSampleSize() {
+      return 0; // not fixed size.
+    }
+
+    @Override
+    public int getSampleSizeAt(int i) {
+      if (fieldSize == 8) {
+        return atom.data.readUnsignedByte();
+      } else if (fieldSize == 16) {
+        return atom.data.readUnsignedShort();
+      } else {
+        // The field size is 4.
+        final boolean isUpperBits = i % 2 == 0;
+        atom.data.setPosition(FIRST_DATA_POSITION + i / 2);
```

Comment:
It feels preferable to avoid having to set the position every time. Can you just store the other 4 bits somewhere for next time?